target/linux/qualcommax/ipq50xx/base-files/lib/upgrade/platform.sh

```diff
@@ -69,8 +71,100 @@ platform_check_image() {
 	return 0;
 }
 
+glinet_validate_firmware() {
+	local type=$(identify_magic_long $(get_magic_long $1))
+	if [ "$type" = "unknown 73797375" ]; then type="sys"; fi
+
+	echo "$type"
+}
+
+glinet_do_fit_upgrade() {
+	echo -n "fit: Extract [ FIT IMAGE ] -x-x-> [ ubi.bin ] ... "
+	local ubi=/tmp/ubi.bin
+	local part=$(dumpimage -l /tmp/firmware.bin | grep -o "Image [0-9] (ubi)" | cut -f2 -d" ")
+
+	local ubibin=$( dumpimage -T flat_dt -p ${part} -o "$ubi"  $1 )
+	if [ -s "$ubi" ]; then
+		echo "[ OK ]"
+		local ubiMd5=$(cat $ubi | md5sum | cut -f1 -d" ")
+		local ubi_size=$( cat "$ubi" | wc -c )
+		echo -n "fit-copy: [ ubi.bin ] -c-c-> [ firmware.bin ] ... "
+		mv "$ubi" "$1"
+		local firmMd5=$(cat $1 | md5sum | cut -f1 -d" ")
+		local firm_size=$( cat $1 | wc -c )
+		if [ ${firm_size} -eq ${ubi_size} ] && [ "$ubiMd5" = "$firmMd5" ]; then
+			echo "[ OK ]"
+			echo "fit-copy: Copied "$firm_size" / "$ubi_size" bytes into firmware.bin"
+			echo "fit-copy: MD5 CHECK: [ OK ]"
+			echo "$ubiMd5 <=> $firmMd5"
+			echo "fit: Successfully Extracted UBI from FIT IMAGE"
+			echo "fit: Proceeding with sysupgrade .."
+			nand_do_upgrade "$1"
+			return
+		fi
+		echo "[ FAILED ] !!"
+		echo "fit-copy: Copied "$firm_size" / "$ubi_size" bytes into firmware.bin"
+		echo "ERROR: Failed to Copy UBI into firmware.bin !!"
+		echo "fit: Terminating sysupgrade .."
+		exit 1
+	fi
+	echo "[ FAILED ] !!"
+	echo "fit-extract: Failed to Create Temp File ubi.bin !!"
+	echo "ERROR: Failed to Extract UBI from FIT IMAGE !!"
+	echo "fit: Terminating sysupgrade .."
+	exit 1
+}
+
+glinet_do_ubi_upgrade() {
+	echo -n "ubi: Removing Metadata Trailer from the UBI Volume ... "
+
+	local metadata=$(fwtool -q -t -i /dev/null "$1")
+	if [ -s $1 ]; then
+		echo "[ OK ]"
+		echo "ubi-meta: Successfully Removed Metadata from UBI Volume"
+		echo "ubi: Proceeding with sysupgrade .."
+		nand_do_upgrade "$1"
+		return
+	fi
+	echo "[ FAILED ] !!"
+	echo "ubi-meta: Cannot remove Metadata, the Files is Empty !!"
+	echo "ERROR: Failed to Remove Metadata Trailer from UBI Volume !!"
+	echo "ubi: Terminating sysupgrade .."
+	exit 1
+}
+
 platform_do_upgrade() {
 	case "$(board_name)" in
+	glinet,gl-b3000)
+		CI_UBIPART="rootfs"
+		echo -n "Validating Firmware ... "
+		case $(glinet_validate_firmware $1) in
```

Comment:
All `$1` should be quoted here:
```suggestion
		case $(glinet_validate_firmware "$1") in
```